cmd/cstorvolume-claim/cvc/controller_base.go

```diff
@@ -0,0 +1,314 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cvc
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	clientset "github.com/openebs/maya/pkg/client/generated/clientset/versioned"
+	openebsScheme "github.com/openebs/maya/pkg/client/generated/clientset/versioned/scheme"
+	informers "github.com/openebs/maya/pkg/client/generated/informers/externalversions"
+	listers "github.com/openebs/maya/pkg/client/generated/listers/openebs.io/v1alpha1"
+	ndmclientset "github.com/openebs/maya/pkg/client/generated/openebs.io/ndm/v1alpha1/clientset/internalclientset"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const controllerAgentName = "cstorvolumeclaim-controller"
+
+// Controller is the controller implementation for SPC resources
+type Controller struct {
+	// kubeclientset is a standard kubernetes clientset
+	kubeclientset kubernetes.Interface
+
+	// clientset is a openebs custom resource package generated for custom API group.
+	clientset clientset.Interface
+
+	// ndmclientset is a ndm custom resource package generated for custom API group.
+	ndmclientset ndmclientset.Interface
+
+	cvcLister listers.CStorVolumeClaimLister
+	cvLister  listers.CStorVolumeLister
+	cspLister listers.CStorPoolLister
+	// cvcSynced is used for caches sync to get populated
+	cvcSynced cache.InformerSynced
+
+	// Store is a generic object storage interface. Reflector knows how to watch a server
+	// and update a store. A generic store is provided, which allows Reflector to be used
+	// as a local caching system, and an LRU store, which allows Reflector to work like a
+	// queue of items yet to be processed.
+	cvcStore cache.Store
+
+	// workqueue is a rate limited work queue. This is used to queue work to be
+	// processed instead of performing it as soon as a change happens. This
+	// means we can ensure we only process a fixed amount of resources at a
+	// time, and makes it easy to ensure we are never processing the same item
+	// simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// ControllerBuilder is the builder object for controller.
+type ControllerBuilder struct {
+	Controller *Controller
+}
+
+// NewControllerBuilder returns an empty instance of controller builder.
+func NewControllerBuilder() *ControllerBuilder {
```

Comment:
should we make it as singleton @AmitKumarDas ?

Reply:
I believe this is a typical case of non singleton instance. Since the builder instance is meant to be used by multiple threads and is mutated, it should be non singleton.

Reply:
We might.. it makes sense..